driver-core/src/main/com/mongodb/internal/async/function/RetryState.java

```diff
@@ -150,19 +191,36 @@ private void doAdvanceOrThrow(final Throwable attemptException,
         if (onlyRuntimeExceptions) {
             assertTrue(isRuntime(attemptException));
         }
-        assertTrue(!isFirstAttempt() || exception == null);
-        Throwable newlyChosenException = callOnAttemptFailureOperator(exception, attemptException, onlyRuntimeExceptions, onAttemptFailureOperator);
-        if (isLastAttempt()) {
-            exception = newlyChosenException;
-            throw exception;
+        assertTrue(!isFirstAttempt() || previouslyChosenException == null);
+        Throwable newlyChosenException = callOnAttemptFailureOperator(previouslyChosenException, attemptException, onlyRuntimeExceptions, onAttemptFailureOperator);
```

Comment:
I believe this `previouslyChosenException` and the exception-wrapping logic below should be moved outside of RetryState, and into (I think?) the retryPredicate.

There might be a bug arising from this. This async retry API is intended to be general, and is used, for example, in OIDC retries. The exceptions thrown in that particular loop (whether auth exceptions, timeout exceptions, or otherwise) should either force a retry, or be thrown as-is. They should not interact with CSOT specific logic, like storage of previously chosen exceptions, or exception wrapping, as appears to be the case, even when an operation timeout is not set up in RetryState. This might (or might not) cause double-wrapping.

Reply:
Currently, the existing OIDC tests are passing without CSOT enabled, which suggests that we should not expect a regression. However, to address the concerns related to CSOT and its interactions with OIDC retries, we can further investigate and make necessary adjustments as part of the ongoing work in [JAVA-5337](https://jira.mongodb.org/browse/JAVA-5337).